runtime/commonMain/src/kotlinx/serialization/modules/SerialModule.kt

```diff
@@ -1,60 +1,124 @@
      */
-    public fun <T : Any> getPolymorphic(baseClass: KClass<T>, serializedClassName: String): KSerializer<out T>?
+    public abstract fun <T : Any> getPolymorphic(baseClass: KClass<in T>, serializedClassName: String): DeserializationStrategy<out T>?
 
     /**
      * Copies contents of this module to the given [collector].
      */
-    public fun dumpTo(collector: SerialModuleCollector)
+    public abstract fun dumpTo(collector: SerialModuleCollector)
 }
 
 /**
  * A [SerialModule] which is empty and always returns `null`.
  */
-public object EmptyModule : SerialModule {
+public object EmptyModule : SerialModule() {
     public override fun <T : Any> getContextual(kclass: KClass<T>): KSerializer<T>? = null
-    public override fun <T : Any> getPolymorphic(baseClass: KClass<T>, value: T): KSerializer<out T>? = null
+    public override fun <T : Any> getPolymorphic(baseClass: KClass<in T>, value: T): SerializationStrategy<T>? = null
     public override fun <T : Any> getPolymorphic(
-        baseClass: KClass<T>,
+        baseClass: KClass<in T>,
         serializedClassName: String
-    ): KSerializer<out T>? = null
+    ): DeserializationStrategy<out T>? = null
 
     public override fun dumpTo(collector: SerialModuleCollector) = Unit
 }
+
+internal typealias PolymorphicProvider<Base> = (className: String) -> DeserializationStrategy<out Base>?
+
+/**
+ * A default implementation of [SerialModule]
+ * which uses hash maps to store serializers associated with KClasses.
+ */
+@Suppress("UNCHECKED_CAST")
+internal class SerialModuleImpl(
+    private val class2Serializer: Map<KClass<*>, KSerializer<*>>,
+    private val polyBase2Serializers: Map<KClass<*>, Map<KClass<*>, KSerializer<*>>>,
+    private val polyBase2NamedSerializers: Map<KClass<*>, Map<String, KSerializer<*>>>,
+    private val polyBase2DefaultProvider: Map<KClass<*>, PolymorphicProvider<*>>
+) : SerialModule() {
+
+    override fun <T : Any> getPolymorphic(baseClass: KClass<in T>, value: T): SerializationStrategy<T>? {
+        if (!value.isInstanceOf(baseClass)) return null
+        val custom = polyBase2Serializers[baseClass]?.get(value::class) as? SerializationStrategy<T>
+        if (custom != null) return custom
+        if (baseClass == Any::class) {
+            val serializer = StandardSubtypesOfAny.getSubclassSerializer(value)
+            return serializer as? SerializationStrategy<T>
+        }
+        return null
+    }
+
+    override fun <T : Any> getPolymorphic(baseClass: KClass<in T>, serializedClassName: String): DeserializationStrategy<out T>? {
+        // Any subtypes
+        val standardPolymorphic =
+            if (baseClass == Any::class) StandardSubtypesOfAny.getDefaultDeserializer(serializedClassName)
+            else null
+        if (standardPolymorphic != null) return standardPolymorphic as KSerializer<out T>
+        // Registered
+        val registered = polyBase2NamedSerializers[baseClass]?.get(serializedClassName) as? KSerializer<out T>
```

Comment:
Shouldn't we place custom registered serializers before the standard any subtypes, as in `getPolymorphic(baseClass, value)`?

Reply:
It was here before my change, but it indeed looks like a bug. Fixed